lxd/instance/drivers/driver_qemu.go

```diff
@@ -7450,54 +7460,62 @@ func (d *qemu) getVsockID() (uint32, error) {
 	return uint32(vsockID), nil
 }
 
-// freeVsockID returns true if the given vsockID is not yet acquired.
-func (d *qemu) freeVsockID(vsockID uint32) bool {
-	c, err := lxdvsock.Dial(vsockID, shared.HTTPSDefaultPort)
-	if err != nil {
-		var unixErrno unix.Errno
+// acquireVsockID tries to occupy the given vsock Context ID.
+// If the ID is free it returns the corresponding file handle.
+func (d *qemu) acquireVsockID(vsockID uint32) (*os.File, error) {
+	vsockIDAcquistionLock.Lock()
+	defer vsockIDAcquistionLock.Unlock()
 
-		if !errors.As(err, &unixErrno) {
-			return false
-		}
+	f, err := os.OpenFile("/dev/vhost-vsock", syscall.O_RDWR, 0)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to open vhost socket: %w", err)
+	}
 
-		if unixErrno != unix.ENODEV {
-			// Skip the vsockID if another syscall error was encountered.
-			return false
+	// 0x4008AF60 = VHOST_VSOCK_SET_GUEST_CID = _IOW(VHOST_VIRTIO, 0x60, __u64)
+	_, _, errno := unix.Syscall(unix.SYS_IOCTL, f.Fd(), 0x4008AF60, uintptr(unsafe.Pointer(&vsockID)))
```

Comment:
Sure I'll try that instead.

Reply:
The only one that looks reasonable is `unix.IoctlSetPointerInt` but it converts the value to `int32` which leads the syscall to always return `invalid argument`.

Reply:
I found [this](https://github.com/golang/go/issues/45585) interesting thread why there are no wrappers for each type. Since we cannot use `unix.IoctlSetPointerInt` due to the wrong int type (int32), the only other thing I could imagine is we create our own wrapper. What do you think?

Reply:
OK, did you try IoctlSetInt?

Reply:
But if no good, then its fine as it is.

Reply:
Jup, needs to be a pointer according to https://www.man7.org/linux/man-pages/man7/vsock.7.html

Reply:
I cant see a reference to VHOST_VSOCK_SET_GUEST_CID in there?

Reply:
But yeah does seem to be a pointer generally.

Reply:
You can find the reference to VHOST_VSOCK_SET_GUEST_CID [here](https://github.com/torvalds/linux/blob/master/include/uapi/linux/vhost.h). From combining all those infos I suspect that you just need to provide pointers when interacting with the vsock.

Reply:
Just got hit by this myself and pushed a patch and comment. The `vsockID` argument to the syscall has to be of type `int`. Before it was set to `uint32` through the function argument. Somehow this worked without any errors.